File: src/chipschallenge/blockreactions/IceCornerTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Inventory.Boots;
import chipschallenge.Move.Moves;

public class IceCornerTo extends BlockReaction {

    private IceCornerTo() {
    }
    private static IceCornerTo mInstance = null;

    public static synchronized IceCornerTo getInstance() {
        if (mInstance == null) {
            mInstance = new IceCornerTo();
        }
        return mInstance;
    }

    @Override
    public void react(Block moving, Block standing) throws BlockContainerFullException {
        // Nothin
    }

    @Override
    public boolean canMove(Block moving, Block standing) {
        Moves m = standing.getFacing();
        switch (moving.getFacing()) {
            case UP:
                return m == Moves.UP || m == Moves.LEFT;
            case DOWN:
                return m == Moves.DOWN || m == Moves.RIGHT;
            case LEFT:
                return m == Moves.UP || m == Moves.RIGHT;
            case RIGHT:
                return m == Moves.LEFT || m == Moves.DOWN;
        }
        return false;
    }

    @Override
    public Moves causesSlip(Block moving, Block standing) {
        if (moving.isChip() && hasBoots(Boots.ICESKATES)) {
            return null;
        } else {
            Moves m = standing.getFacing();
            Moves force = null;
            switch (moving.getFacing()) {
                case UP:
                    force = m == Moves.UP ? Moves.RIGHT : Moves.LEFT;
                    break;
                case DOWN:
                    force = m == Moves.DOWN ? Moves.LEFT : Moves.RIGHT;
                    break;
                case LEFT:
                    force = m == Moves.UP ? Moves.DOWN : Moves.UP;
                    break;
                case RIGHT:
                    force = m == Moves.LEFT ? Moves.DOWN : Moves.UP;
                    break;
            }
            return force;
        }
    }
}

File: ChipsChallenge.Shared/Blockreactions/IceCornerTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Boots = Inventory.Boots;
    using Moves = Move.Moves;

    public class IceCornerTo : BlockReaction
    {
        private IceCornerTo()
        {
        }
        private static IceCornerTo instance;

        public static IceCornerTo Instance
        {
            get
            {
                lock (typeof(IceCornerTo))
                {
                    return instance ?? (instance = new IceCornerTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
        }

        public override bool canMove(Block moving, Block standing)
        {
            Moves m = standing.Facing;
            switch (moving.Facing)
            {
                case Moves.UP:
                    return m == Moves.UP || m == Moves.LEFT;
                case Moves.DOWN:
                    return m == Moves.DOWN || m == Moves.RIGHT;
                case Moves.LEFT:
                    return m == Moves.UP || m == Moves.RIGHT;
                case Moves.RIGHT:
                    return m == Moves.LEFT || m == Moves.DOWN;
            }
            return false;
        }

        public override Moves? CausesSlip(Block moving, Block standing)
        {
            if (moving.Chip && HasBoots(Boots.ICESKATES))
            {
                return null;
            }
            else
            {
                Moves m = standing.Facing;
                Moves? force = null;
                switch (moving.Facing)
                {
                    case Moves.UP:
                        force = m == Moves.UP ? Moves.RIGHT : Moves.LEFT;
                        break;
                    case Moves.DOWN:
                        force = m == Moves.DOWN ? Moves.LEFT : Moves.RIGHT;
                        break;
                    case Moves.LEFT:
                        force = m == Moves.UP ? Moves.DOWN : Moves.UP;
                        break;
                    case Moves.RIGHT:
                        force = m == Moves.LEFT ? Moves.DOWN : Moves.UP;
                        break;
                }

                return force;
            }
        }
    }
}